File: src/main/java/calculator/Number.java
package calculator;


public class Number {
    public static final Number ZERO = new Number(0);
    private static final NumberValidator numberValidator = NumberValidator.getInstance();

    private final int value;

    private Number(int value) {  // private 생성자
        this.value = value;
    }

    public static Number from(String input) {
        numberValidator.validate(input);
        return new Number(Integer.parseInt(input));
    }

    public Number add(Number other) {
        return new Number(value + other.value);
    }

    public int value() {
        return value;
    }



}

File: src/main/java/calculator/Number.kt
package calculator

import calculator.exception.NumberFormatException

@JvmInline
value class Number(
    val value: Int,
) {
    constructor(value: String) : this(
        value.toIntOrNull() ?: throw NumberFormatException("Invalid number format: $value")
    )

    init {
        require(value >= ZERO.value) { "음수는 지원하지 않습니다" }
    }

    operator fun plus(number: Number) = Number(Math.addExact(value, number.value))

    companion object {
        val ZERO: Number = Number(0)
    }
}
